crypto/key_manager.go

```diff
@@ -82,140 +59,84 @@ func (k PEMKeyManager) HashAlgorithm() crypto.Hash {
 	return crypto.SHA256
 }
 
-// LoadPrivateKey loads a private key from a PEM encoded string, decrypting it if necessary
-func (k *PEMKeyManager) LoadPrivateKey(pemEncodedKey, password string) error {
-	block, rest := pem.Decode([]byte(pemEncodedKey))
-	if len(rest) > 0 {
-		return errors.New("extra data found after PEM decoding")
-	}
-
-	der := block.Bytes
-	if password != "" {
-		pwdDer, err := x509.DecryptPEMBlock(block, []byte(password))
-		if err != nil {
-			return err
-		}
-		der = pwdDer
-	}
-
-	key, algo, err := parsePrivateKey(der)
-	if err != nil {
-		return err
-	}
-
-	k.serverPrivateKey = key
-	k.signatureAlgorithm = algo
-	signer, err := k.Signer()
-	if err != nil {
-		return err
-	}
-	k.serverPublicKey = signer.Public()
-	return nil
+// Signer returns a signer based on our private key.
+func (k PEMKeyManager) Signer() crypto.Signer {
+	return k.signer
 }
 
-// LoadPublicKey loads a public key from a PEM encoded string.
-func (k *PEMKeyManager) LoadPublicKey(pemEncodedKey string) error {
-	publicBlock, rest := pem.Decode([]byte(pemEncodedKey))
-	if publicBlock == nil {
-		return errors.New("could not decode PEM for public key")
-	}
-	if len(rest) > 0 {
-		return errors.New("extra data found after PEM key decoded")
-	}
-
-	k.rawPublicKey = publicBlock.Bytes
-
-	parsedKey, err := x509.ParsePKIXPublicKey(publicBlock.Bytes)
-	if err != nil {
-		return errors.New("unable to parse public key")
-	}
+// PublicKey returns the public key corresponding to the private key.
+func (k PEMKeyManager) PublicKey() crypto.PublicKey {
+	return k.signer.Public()
+}
 
-	k.serverPublicKey = parsedKey
-	return nil
+// NewFromPrivateKey creates PrivateKeyManager using a private key.
+func NewFromPrivateKey(key crypto.PrivateKey) (PrivateKeyManager, error) {
+	var signer crypto.Signer
+	var sigAlgo sigpb.DigitallySigned_SignatureAlgorithm
+
+	switch key.(type) {
```

Comment:
Why not `switch key := key.(type) {`? That would eliminate the need for type assertions on lines 79 and 82.